crates/ln-dlc-node/src/node/mod.rs

```diff
@@ -538,15 +492,79 @@ where
             storage: node_storage,
             fee_rate_estimator,
             ldk_config,
-            _background_processor_handle: background_processor_handle,
-            _connection_manager_handle: connection_manager_handle,
-            _broadcast_node_announcement_handle: broadcast_node_announcement_handle,
-            _sub_channel_manager_periodic_check_handle: sub_channel_manager_periodic_check_handle,
             network_graph,
             settings,
+            listen_address,
+            gossip_sync,
+            persister,
+            alias: alias.to_string(),
+            announcement_addresses,
+            scorer,
+            esplora_server_url,
         })
     }
 
+    /// Starts the background handles - if the returned handles are dropped, the
+    /// background tasks are stopped.
+    pub fn start(&self, event_sender: Option<watch::Sender<Option<Event>>>) -> Result<RunningNode> {
+        let mut handles = vec![spawn_connection_management(
+            self.peer_manager.clone(),
+            self.listen_address,
+        )];
+
+        // TODO: Allow passing in event handler from outside
+        // Maybe there's a 'friend' class or something in Rust?
+        let event_handler = EventHandler::new(
+            self.channel_manager.clone(),
+            self.sub_channel_manager.clone(),
+            self.wallet.clone(),
+            self.network_graph.clone(),
+            self.keys_manager.clone(),
+            self.storage.clone(),
+            self.fake_channel_payments.clone(),
+            Arc::new(Mutex::new(HashMap::new())),
+            self.peer_manager.clone(),
+            self.fee_rate_estimator.clone(),
+            event_sender,
+            self.ldk_config.clone(),
+        );
+
+        handles.push(spawn_background_processor(
+            self.peer_manager.clone(),
+            self.channel_manager.clone(),
+            self.chain_monitor.clone(),
+            self.logger.clone(),
+            self.persister.clone(),
+            event_handler,
+            self.gossip_sync.clone(),
+            self.scorer.clone(),
+        ));
```

Comment:
🤔 I still think the simplest thing to maintain will be to do this separately for each crate: `coordinator`, `app`, `maker`.

We can reuse individual event handlers (and other functions) if we see common patterns, but in general it's a bit of an illusion that the `ln-dlc-node` is a good idea as currently designed.

It's already very apparent that the shared event handler is bad. But there's quite a few other things that are difficult to do because of this crate being used in several places. Some examples:

- The app and the coordinator have different requirements when syncing the wallet. One might want to do it periodically in the background. The other likely wants to do it after certain events: app start-up, user input, transaction has just been published.

- Different persistence strategies. Something running on the phone might want to approach persisting data differently to a daemon that is constantly running in a dedicated server. We have no easy way to diverge the persistence behaviour between app and coordinator with the current design.

- App shouldn't send node announcements because it has no public channels!

The `ln-dlc-node` crate was convenient at first, but I think it has overstayed its welcome. The event handler is the main pain point, but I think we need to move away from `ln_dlc_node::node::Node` as fast as possible.

---

This is all to say that I'm not convinced that letting this crate orchestrate things whilst we pass things from the outside is such a good idea, because we already know that there are certain things that we want to orchestrate differently among consumers.

Reply:
> 🤔 I still think the simplest thing to maintain will be to do this separately for each crate: `coordinator`, `app`, `maker`.
> 
> We can reuse individual event handlers (and other functions) if we see common patterns, but in general it's a bit of an illusion that the `ln-dlc-node` is a good idea as originally conceived.
> 
> It's already very apparent that the shared event handler is bad. But there's quite a few other things that are difficult to do because of this crate being used in several places. Some examples:
> 
> * The app and the coordinator have different requirements when syncing the wallet. One might want to do it periodically in the background. The other likely wants to do it after certain events: app start-up, user input, transaction has just been published.
> * Different persistence strategies. Something running on the phone might want to approach persisting data differently to a daemon that is constantly running in a dedicated server. We have no easy way to diverge the persistence behaviour between app and coordinator with the current design.
> * App shouldn't send node announcements because it has no public channels!
> 
> The `ln-dlc-node` crate was convenient at first, but I think it has overstayed its welcome. The event handler is the main pain point, but I think we need to move away from `ln_dlc_node::node::Node` as fast as possible.
> 
> This is all to say that I'm not convinced that letting this crate orchestrate things whilst we pass things from the outside is such a good idea, because we already know that there are certain things that we want to orchestrate differently among consumers.

Eventually yes, one can split the starting behaviour into different ways.  I don't want to go to far because of the maintenance burden and the fact that this is just so much refactoring to do that one just has to do it in steps. 

Note that this might be really simply done by having a few different policies passed in the constructor. 

I'm about to send an update that passes in Event Handler from the outside, perhaps wait for that before doing another pass at the review :)

Reply:
Perhaps have a look at #1062 .

I'm confident that we can achieve all the requirements you've enlisted above without getting into a world of pain with maintenance burden. - the constructor without starting any tasks still has ~180 lines (!!!)

1) Node will have different settings passed in from outside, including different persistence requirements
2) custom event handlers support
3) running the node can be completely customised